src/test/java/org/jetbrains/kotlinx/lincheck/test/verifier/nesting_safe_recoverable_linearizability/set/NRLSet.java

```diff
@@ -0,0 +1,234 @@
+package org.jetbrains.kotlinx.lincheck.test.verifier.nesting_safe_recoverable_linearizability.set;
+
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicMarkableReference;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @see  <a href="https://arxiv.org/pdf/1905.13600.pdf">Tracking in Order to Recover: Recoverable Lock-Free Data Structures</a>
+ */
+public class NRLSet<T extends Comparable<T>> {
```

Comment:
It would probably be better to write implementations in Kotlin since it is simply more readable